library/psa_crypto.c

```diff
@@ -3698,9 +3698,9 @@ psa_status_t psa_sign_hash( psa_key_handle_t handle,
 #if defined(MBEDTLS_ECP_C)
     if( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
     {
-#if defined(MBEDTLS_ECDSA_C)
+#if defined(PSA_WANT_ALG_ECDSA)
         if(
```

Comment:
This block and its condition are a bit subtle due to how PSA and Mbed TLS configure the availability of randomized vs deterministic ECDSA differently.

The block needs to be included if either `MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA` or `MBEDTLS_PSA_BUILTIN_ALG_ECDSA` is enabled.

Mbed TLS can be configured to support both deterministic and randomized ECDSA, randomized ECDSA only, or neither, but not deterministic ECDSA only. So we don't actually have to address the case where `MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA` is enabled but `MBEDTLS_PSA_BUILTIN_ALG_ECDSA` is disabled: this would trigger a compile-time configuration verification error. It would be more robust to address it in case the Mbed TLS configuration changes (this might happen in Mbed TLS 3.0).

In hindsight it would be more readable to define an auxiliary macro `IS_SUPPORTED_ECDSA(alg)` that is an alias for one of `PSA_ALG_IS_ECDSA`, `PSA_ALG_IS_RANDOMIZED_ECDSA`, `PSA_ALG_IS_DETERMINISTIC_ECDSA` or `0` depending on the configuration, even though this macro would only be used in this one place.

Reply:
I fixed the inclusion of the block, but did not go as far to do the IS_SUPPORTED_ECDSA change. That seemed like it was more extensive and would likely require additional test changes to make sure it is working properly for all instances.